src/validation.h

```diff
@@ -757,6 +762,111 @@ bool InvalidateBlock(BlockValidationState& state, const CChainParams& chainparam
 
+/**
+ * Provides an interface for creating and interacting with multiple
+ * chainstates, as well as a means to initialize chainstates from a
+ * UTXO snapshot. Managed chainstates can be maintained at different
+ * heights simultaneously.
+ *
+ * This class provides abstractions that allow the retrieval of the current
+ * most-work chainstate ("Active") as well as chainstates which may be in
+ * background use to validate UTXO snapshots.
+ *
+ * A few quick definitions:
+ *
+ * *IBD chainstate*: a chainstate populated via initial block download.
+ *
+ * *Snapshot chainstate*: a chainstate populated by loading in an
+ *    assumed-valid UTXO snapshot.
+ *
+ * *Active chainstate*: the chainstate containing the current most-work
+ *    chain. Consulted by most parts of the system (net_processing,
+ *    wallet) as a reflection of the current chain and UTXO set.
+ *    This may either be an IBD chainstate or a snapshot chainstate.
+ *
+ * *Background validation chainstate*: an IBD chainstate for which the
+ *    IBD process is happening in the background while use of the
+ *    active chainstate allows the rest of the system to function.
+ *
+ * *Validated chainstate*: a chainstate whose validity is not assumed.
+ *    This could be an IBD chainstate, or a snapshot chainstate for
+ *    which background validation (up to the base of the snapshot)
+ *    has completed.
+ */
+class ChainstateManager
+{
+private:
+    //! The chainstate used under normal operation (i.e. "regular" IBD) or,
+    //! if a snapshot is in use, for background validation. Its contents will
+    //! be freed when background validation of the snapshot has completed.
+    std::unique_ptr<CChainState> m_ibd_chainstate;
+
+    //! A chainstate initialized on the basis of a UTXO snapshot. If this is
+    //! non-null, it is always our active chainstate unless proven invalid.
+    std::unique_ptr<CChainState> m_snapshot_chainstate;
+
+    //! Points to either the ibd or snapshot chainstate; indicates our
+    //! most-work chain.
+    CChainState* m_active_chainstate;
+
+    //! The base blockhash of the active UTXO snapshot (if one is being used)
+    uint256 m_snapshot_blockhash{};
+
+    //! If true, the assumed-valid chainstate has been fully validated
+    //! by the background validation chainstate.
+    bool m_snapshot_validated{false};
+
+    // For access to m_active_chainstate.
+    friend CChain& ChainActive();
+
+public:
+    //! Instantiate a new chainstate and assign it based upon whether it is
+    //! from a snapshot.
+    //!
+    //! @param[in] activate   If true, make this new chainstate the active one.
+    //! @param[in] snapshot_blockhash   If given, signify that this chainstate
+    //!                                 is based on a snapshot.
+    CChainState& InitializeChainstate(
+        bool activate = true, const uint256& snapshot_blockhash = uint256());
+
+    //! Get all chainstates currently being used.
+    std::vector<CChainState*> GetAll();
```

Comment:
Agree. Same question for other methods.

Reply:
Thanks for raising this point. Initially, I had all pointers in the ChainstateManager protected by a private `m_cs` lock. I later found that there were irreconcilable lock inversions that happened during initialization - because a function like `ChainActive()` would necessarily have to acquire `ChainstateManager.m_cs`, there were many tricky inversions that happened between that lock and others like `cs_main` since ChainActive() is used so pervasively throughout the code. 

As a result, I removed the chainman-specific lock and replaced it with `cs_main` in a later commit (https://github.com/bitcoin/bitcoin/pull/15606/commits/f1fa917e2081dd8d1847cbb23a0eae5a2cded2e5). Because ChainActive() isn't formally annotated as being protected by cs_main, and use of ChainstateManager always happens while under the protection of cs_main, I figured I could forgo adding those annotations now. But now that the point has been raised, I think it's probably worth addressing in this PR. 

So I'll add the `cs_main` annotations to the relevant ChainManager parts. Though of course this will entail annotating ChainActive() and there will come with many other (mostly trivial) changes.

Reply:
Also, it's worth noting that the comment you quote is out of date (my fault!). We don't actually free the contents of the background validation chainstate *immediately* when validation completes - [we do it during shutdown](https://github.com/bitcoin/bitcoin/pull/15606/commits/8d6e0c34c32c5b95c126a963b16b5aebf707acbd#diff-24efdb00bfbe56b140fb006b562cc70bR5653-R5679) when `g_chainman.Reset()` is called. 

The reason I don't do it immediately after validation is complete is so that we can be sure that no other parts of the code are still making reference to that pointer. This also relates to your question below about `IsSnapshotValidated()` - the extra piece of state (`m_snapshot_validated`) is used to track whether or not we need to do this cleanup.